✨(app) create a redis-sentinel application

lunika opened the pull request:
## Purpose

we choose to use the redis-operator
(https://github.com/spotahome/redis-operator) to deploy a redis
configured with sentinel. Only one OpenShift object is used and deployed as a
Service.


## Proposal

- [x] implement the redis-sentinel app

lunika committed:
✨(app) create a redis-sentinel application

we choose to use the redis-operator
(https://github.com/spotahome/redis-operator) to deploy a redis
configured with sentinel. Only one OpenShift object is used and
deployed as a Service.

jmaupetit commented on apps/redis-sentinel/vars/settings.yml:
```diff
@@ -0,0 +1 @@
+is_blue_green_compatible: False
```
```suggestion
is_blue_green_compatible: false
```

jmaupetit commented on apps/redis-sentinel/templates/services/app/svc.yml.j2:
```diff
@@ -0,0 +1,55 @@
+{# 
+This template is not relative to an OpenShift Service object but to a RedisFailover one.
+RedisFailover is part of redis-operator (https://github.com/spotahome/redis-operator) ans must be
+deploy at the same time of other OpenShift Service, this is why we decided to created in this file.
```
Can you indent this comment in the block?

jmaupetit commented on apps/redis-sentinel/templates/services/app/svc.yml.j2:
```diff
@@ -0,0 +1,55 @@
+{# 
+This template is not relative to an OpenShift Service object but to a RedisFailover one.
```
```suggestion
Disclaimer: this template is not relative to an OpenShift Service object but to a RedisFailover object.
```

jmaupetit commented on apps/redis-sentinel/templates/services/app/svc.yml.j2:
```diff
@@ -0,0 +1,55 @@
+{# 
+This template is not relative to an OpenShift Service object but to a RedisFailover one.
+RedisFailover is part of redis-operator (https://github.com/spotahome/redis-operator) ans must be
```
```suggestion
RedisFailover is part of the redis-operator (https://github.com/spotahome/redis-operator) and must be
```

jmaupetit commented on apps/redis-sentinel/templates/services/app/svc.yml.j2:
```diff
@@ -0,0 +1,55 @@
+    app: redis-sentinel
+    service: app
+    version: RedisFailover
+  # name of the service should be host name in edxapp configuration prefixed by rfs- : rfs-redis-sentinel
```
```suggestion
  # Nota bene: the name of the service should be the host name in edxapp configuration,
  # prefixed by "rfs-", e.g. "rfs-redis-sentinel"
```

jmaupetit commented on apps/redis-sentinel/vars/all/main.yml:
```diff
@@ -0,0 +1,26 @@
+# This setting can be found in the project annotation
```
```suggestion
# The following settings can be found in the project's annotation.
```

jmaupetit commented on apps/redis-sentinel/vars/all/main.yml:
```diff
@@ -0,0 +1,26 @@
+# This setting can be found in the project annotation
+# to retrieve them run oc project {{ project_name }} -o yaml
```
```suggestion
# To retrieve them run the following command:
#   $ oc project {{ project_name }} -o yaml
```

jmaupetit commented on apps/redis-sentinel/vars/all/main.yml:
```diff
@@ -0,0 +1,26 @@
+# This setting can be found in the project annotation
+# to retrieve them run oc project {{ project_name }} -o yaml
+# then copy the value of the annotation openshift.io/sa.scc.supplemental-groups
```
```suggestion
# then copy values of the `openshift.io/sa.scc.supplemental-groups` annotation
```

jmaupetit commented on apps/redis-sentinel/vars/all/main.yml:
```diff
@@ -0,0 +1,26 @@
+redis_sentinel_redis_replicas: 3
+redis_sentinel_redis_data_volume_size: 2Gi
+# If you need custom redis configuration you should add
+# them in this list. One configuration is a new entry
```
```suggestion
# them in this list with one configuration per item, e.g.:
```

jmaupetit commented on apps/redis-sentinel/vars/all/main.yml:
```diff
@@ -0,0 +1,26 @@
+# Number of sentinel replicas 
+redis_sentinel_sentinel_replicas: 3
+# If you need custom sentinel configuration you should add
+# them in this list. One configuration is a new entry
```
```suggestion
# them in this list with one configuration per item, e.g.:
```

jmaupetit reviewed (approved):
Great work!

![](https://media0.giphy.com/media/IUZtGhVO8hZ6w/giphy.gif)

jmaupetit commented:
Even tho, we will need a new PR allowing to use Redis sentinel in `edxapp` (see worker probes).

jmaupetit reviewed (changes requested):
In fact we need to fix `edxapp` workers probes that would be broken if we switch to redis sentinel. See https://github.com/openfun/arnold/blob/master/apps/edxapp/templates/services/cms/_dc_base.yml.j2#L96

lunika commented:
> In fact we need to fix `edxapp` workers probes that would be broken if we switch to redis sentinel. See https://github.com/openfun/arnold/blob/master/apps/edxapp/templates/services/cms/_dc_base.yml.j2#L96

This should be done in an other PR ? This PR just target the creation of a new app

jmaupetit commented:
> This PR just target the creation of a new app

Indeed, but it's unusable as is.

lunika commented:
Ok so the only solution I found for now is to transform the command with this:

```
celery -A lms.celery inspect ping -d celery@{{ celery_worker.name }}.%$(hostname)
```

WDYT @sampaccoud @jmaupetit

sampaccoud commented:
@lunika I would say that if we want the edxapp healthcheck to validate that redis is working it should be done via the edxapp healthcheck endpoint? But I'm not really aware of existing probes :man_shrugging:

lunika commented:
> @lunika I would say that if we want the edxapp healthcheck to validate that redis is working it should be done via the edxapp healthcheck endpoint? But I'm not really aware of existing probes man_shrugging

This probe is not used by the wsgi application but by every worker deployed so the healthcheck endpoint is not usable.

sampaccoud commented:
Then we should add a small management command in edxapp that checks the redis connection (with or without sentinel depending on the settings the same way as Celery or django-redis-session are connecting to redis...). I would make a simple command that writes a value to redis and reads it again... 
Actually it could also be the same endpoint for the wsgi application and the worker pods... WDYT?

jmaupetit commented:
Loading a management command with the `edx`stack is too heavy for a health check. That was our first attempt and it proved to be unusable.

jmaupetit commented:
> Ok so the only solution I found for now is to transform the command with this:
> 
> ```
> celery -A lms.celery inspect ping -d celery@{{ celery_worker.name }}.%$(hostname)
> ```
> 
> WDYT @sampaccoud @jmaupetit

Have you tried it with `time` in a worker pod to evaluate its responsiveness.

lunika commented:
``` bash
$ time celery -A lms.celery inspect ping -d celery@lms-default.%$(hostname)
-> celery@lms-default.%edxapp-lms-default-worker-d-190717-13h37m29s-2-sf9qm: OK
        pong

real	0m1.211s
user	0m0.992s
sys	0m0.620s
```

jmaupetit commented:
Did you run it multiple times (at least 3) to check execution time "standard deviation"? IMO 1.2s is not acceptable for such probe.

lunika commented:
> Did you run it multiple times (at least 3) to check execution time "standard deviation"? IMO 1.2s is not acceptable for such probe.

Yes I run it multiple time and this is one the best time I have.

jmaupetit commented:
So it's a dead end.

lunika committed:
🔥(edxapp) remove probes on worker pods

Probes used on workers to test redis with a celery command are not
usable anymore with redis sentinel. To test it we must have to use the
celery module (lme.celery) and it takes to much time to run it.

jmaupetit commented on CHANGELOG.md:
```diff
@@ -14,6 +14,10 @@ Versioning](http://semver.org/spec/v2.0.0.html).
 
+### Removed
+
+- probes on edxapp workers. They are too slow to be run
```
This is not true: the real problem is that existing probes are not compatible with redis sentinel and alternative probes based on Django management commands or Celery using Django settings are too slow to be considered as a relevant alternative. You must add that this is a temporary removal until we implement a new efficient probe.

jmaupetit commented on CHANGELOG.md:
```diff
@@ -8,6 +8,19 @@ Versioning](http://semver.org/spec/v2.0.0.html).
+
+### Removed
+
+- probes on edxapp workers. Existing probes are not compatible with redis sentinel.
```
```suggestion
- Probes on edxapp workers. Existing probes are not compatible with redis sentinel.
```

jmaupetit commented on CHANGELOG.md:
```diff
@@ -8,6 +8,19 @@ Versioning](http://semver.org/spec/v2.0.0.html).
+
+- probes on edxapp workers. Existing probes are not compatible with redis sentinel.
+  To use them with redis sentinel, we must use a Django management command or Celery
+  using Django settings, doing this is too slow to be considered as a relevant alternative.
```
```suggestion
  using Django settings, but doing so is too slow to be considered as a relevant alternative.
```

jmaupetit commented on CHANGELOG.md:
```diff
@@ -8,6 +8,19 @@ Versioning](http://semver.org/spec/v2.0.0.html).
+- probes on edxapp workers. Existing probes are not compatible with redis sentinel.
+  To use them with redis sentinel, we must use a Django management command or Celery
+  using Django settings, doing this is too slow to be considered as a relevant alternative.
+  This removal is temporary, a new solution must be developed soon.
```
```suggestion
  This removal is temporary, a new solution will be developed soon.
```
